src/smartcontracts/typesystem/typeExpressionParser.ts

```diff
@@ -3,83 +3,83 @@ import { Type } from "./types";
-      throw new errors.ErrTypingSystem(`cannot parse type expression: ${expression}. internal json: ${jsoned}.`);
+        try {
+            return jsonHandler.parse(jsoned);
+        } catch (error) {
+            throw new errors.ErrTypingSystem(`cannot parse type expression: ${expression}. internal json: ${jsoned}.`);
+        }
     }
-  }
 
-  /**
-   * Converts a raw type expression to a JSON, parsing-friendly format.
-   * This is a workaround, so that the parser implementation is simpler (thus we actually rely on the JSON parser).
-   *
-   * @param expression a string such as:
-   *
-   * ```
-   *  - Option<List<Address>>
-   *  - VarArgs<MultiArg2<bytes, Address>>
-   *  - MultiResultVec<MultiResult2<Address, u64>
-   * ```
-   */
-  private getJsonedString(expression: string) {
-    let jsoned = "";
+    /**
+     * Converts a raw type expression to a JSON, parsing-friendly format.
+     * This is a workaround, so that the parser implementation is simpler (thus we actually rely on the JSON parser).
+     *
+     * @param expression a string such as:
+     *
+     * ```
+     *  - Option<List<Address>>
+     *  - VarArgs<MultiArg2<bytes, Address>>
+     *  - MultiResultVec<MultiResult2<Address, u64>
+     * ```
+     */
+    private getJsonedString(expression: string) {
+        let jsoned = "";
 
-    for (var i = 0; i < expression.length; i++) {
-      let char = expression.charAt(i);
-      let previousChar = expression.charAt(i - 1);
-      let nextChar = expression.charAt(i + 1);
+        for (var i = 0; i < expression.length; i++) {
+            let char = expression.charAt(i);
+            let previousChar = expression.charAt(i - 1);
+            let nextChar = expression.charAt(i + 1);
 
-      if (char == "<") {
-        jsoned += ": {";
-      } else if (char == ">") {
-        if (previousChar != ">") {
-          jsoned += ": {} }";
-        } else {
-          jsoned += "}";
-        }
-      } else if (char == ",") {
-        if (nextChar == ">") {
-          // Skip superfluous comma
-        } else {
-          jsoned += ": {},";
+            if (char == "<") {
+                jsoned += ": {";
+            } else if (char == ">") {
+                if (previousChar != ">") {
+                    jsoned += ": {} }";
+                } else {
+                    jsoned += "}";
+                }
+            } else if (char == ",") {
+                if (nextChar == ">") {
+                    // Skip superfluous comma
+                } else {
+                    jsoned += ": {},";
+                }
+            } else {
+                jsoned += char;
+            }
         }
-      } else {
-        jsoned += char;
-      }
-    }
 
-    let symbolsRegex = /(:|\{|\}|,|\s)/;
-    let tokens = jsoned.split(symbolsRegex).filter((token) => token);
-    jsoned = tokens.map((token) => (symbolsRegex.test(token) ? token : `"${token}"`)).join("");
+        let symbolsRegex = /(:|\{|\}|,|(?<!utf\-8)\s)/;
```

Comment:
The `utf-8 string` is a single type but because this code was splitting by space, I had to add an exception.
The `(?<!utf\-8)` is a negative look-ahead.
I added a comment for this in the code.

Reply:
I would recommend moving these lines to a method called `splitBySpaceExceptUTF8()`, then adding a few small unit tests just for it.

Reply:
I can't neatly split this into a separate function (since the `symbolsRegex` is used in 2 places, and bringing both lines into the function would bring the rest of the code in as well, because of the re-used `tokens` and `jsoned` ).
However, this function is well tested through the `parse` method. I added a test-case for the `utf-8 string` to the already existing tests in `typeExpressionParser.spec.ts`.